### PR TITLE
add prisma scan

### PR DESCRIPTION
**Pull request by pgvishnuram**
## Details

Add prisma scan as a secondary scan with twistcli that sends results to prisma cloud
## Related Issues

https://github.com/astronomer/issues/issues/5876

### CODE REVIEW

File: .circleci/continue-config.yml.j2 [hunk 1]
```diff
@@ -140,6 +146,37 @@ jobs:
...
+      - run:
+          name: Install twistcli
+          command: |
+            apk add --update-cache --upgrade curl
+            curl -k -u $USER:$PASSWORD --output twistcli $CONSOLE_URL"/api/v1/util/twistcli"
+            chmod +x twistcli
+      - run:
+          name: scan image
+          command: |
+            ./twistcli images scan --address $CONSOLE_URL --user=$USER --password=$PASSWORD --details ap-"<< parameters.directory >>":${CIRCLE_SHA1}
```
Comment (danielhoherd) [inline .circleci/continue-config.yml.j2:179]:
```suggestion
            ./twistcli images scan --address $CONSOLE_URL --user=$USER --password=$PASSWORD --details "ap-<< parameters.directory >>:${CIRCLE_SHA1}"
```

File: .circleci/continue-config.yml [hunk 2]
```diff
@@ -2065,6 +2317,37 @@ jobs:
...
+          name: Load archived Docker image
+          command: |
+            set -x
+            docker load -i /tmp/workspace/ap-<< parameters.directory >>.tar
+            docker images | grep ap-<< parameters.directory >>
+      - run:
+          name: Install twistcli
+          command: |
+            apk add --update-cache --upgrade curl
+            curl -k -u $USER:$PASSWORD --output twistcli $CONSOLE_URL"/api/v1/util/twistcli"
```
Comment (danielhoherd) [inline .circleci/continue-config.yml:2345]:
```suggestion
            curl -k -u "$USER:$PASSWORD" --output twistcli "$CONSOLE_URL/api/v1/util/twistcli"
```

File: .circleci/continue-config.yml.j2 [hunk 3]
```diff
@@ -140,6 +146,37 @@ jobs:
...
+          name: Load archived Docker image
+          command: |
+            set -x
+            docker load -i /tmp/workspace/ap-<< parameters.directory >>.tar
+            docker images | grep ap-<< parameters.directory >>
+      - run:
+          name: Install twistcli
+          command: |
+            apk add --update-cache --upgrade curl
+            curl -k -u $USER:$PASSWORD --output twistcli $CONSOLE_URL"/api/v1/util/twistcli"
```
Comment (danielhoherd) [inline .circleci/continue-config.yml.j2:174]:
```suggestion
            curl -k -u "$USER:$PASSWORD" --output twistcli "${CONSOLE_URL}/api/v1/util/twistcli"
```

**Review by danielhoherd — CHANGES_REQUESTED** [review]
This currently does not error when a vulnerability is found. <https://app.circleci.com/pipelines/github/astronomer/ap-vendor/2156/workflows/57485111-9a92-427c-9691-d3df558e3ef0/jobs/43279/parallel-runs/0/steps/0-105> We need this step to fail if a vulnerability is found unless that vulnerability has been ignored.

**Comment by pgvishnuram** [comment]
I think thats fine daniel since else you will keep ending up on vulnerabilities that we are not responsible for

**Comment by danielhoherd** [comment]
@pgvishnuram 

> vulnerabilities that we are not responsible for

We are responsible for all vulnerabilities in the software we ship. We can't release software that has a vulnerability that we have not acknowledged and added an exception for.

To put it a different way, the CI security scan serves a specific purpose: alerting a developer like you or me that there is a vulnerability in the image. Not failing on this steps means that we will inevitably release software that has a vulnerability in it that we didn't see.

Not failing in CI is a big step down in security confidence. CC @VictorLuc4

**Comment by pgvishnuram** [comment]
@danielhoherd Im merging this and later we can tweak it as per the needs since we have a open ticket

File: bin/twistcli_validator.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import json
+import sys
+import os
+from pathlib import Path
+
+
+ROOT_DIRECTORY = Path(__file__).parent.parent
```
Comment (danielhoherd) [inline bin/twistcli_validator.py:7]:
```suggestion
ROOT_DIRECTORY = [x for x in Path(__file__).resolve().parents if (x / ".git").is_dir()][-1]
```

File: bin/twistcli_validator.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import json
```
Comment (danielhoherd) [inline bin/twistcli_validator.py:1]:
```suggestion
#!/usr/bin/env python3
import json
```

File: bin/twistcli_validator.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+ROOT_DIRECTORY = Path(__file__).parent.parent
+PROJECT_NAME = os.getenv("PROJECT_DIRECTORY")
+SCAN_RESULTS_FILE_PATH = f"{ROOT_DIRECTORY}/{PROJECT_NAME}"
+output = []
+
+with open(f"{SCAN_RESULTS_FILE_PATH}/scan-results.json", "r") as f:
+    data = json.load(f)
+    vulnerabilities = data.get("results", [])[0].get("vulnerabilities", [])
+    output = [vul["id"] for vul in vulnerabilities]
```
Comment (danielhoherd) [inline bin/twistcli_validator.py:15]:
```suggestion
data = json.loads(Path(f"{SCAN_RESULTS_FILE_PATH}/scan-results.json").read_text())
output = [vuln["id"] for vuln in data.get("results", [])[0].get("vulnerabilities", [])]
```

File: bin/twistcli_validator.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+try:
+    with open(f"{SCAN_RESULTS_FILE_PATH}/twistcliignore", "r") as f:
+        ignore_list = [line.strip() for line in f.read().splitlines()]
+        if ignore_list == output:
```
Comment (danielhoherd) [inline bin/twistcli_validator.py:19]: This `output` variable name should be more specific. Maybe `cves` would be good.
```suggestion
        if ignore_list == output:
```

File: bin/twistcli_validator.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+            print("vuln is in ignore list")
+        else:
+            print("New vulnerabilities found.Fix them or add into ignore list")
```
Comment (danielhoherd) [inline bin/twistcli_validator.py:22]: We should print the CVEs that are not ignored. We can do this with `set`:
```
>>> ignore_list = ['CVE-2023-39320', 'CVE-2023-29405', 'CVE-2023-29404', 'CVE-2023-29402']
>>> cves = ['CVE-2023-39320', 'CVE-2023-29405', 'CVE-2023-29404', 'CVE-2023-29402', 'CVE-2023-29403', 'CVE-2023-39533', 'CVE-2023-39322', 'CVE-2023-39321']
>>> set(cves) - set(ignore_list)
{'CVE-2023-39321', 'CVE-2023-39533', 'CVE-2023-39322', 'CVE-2023-29403'}
```

Or with a comprehension:

```
>>> [cve for cve in cves if cve not in ignore_list]
['CVE-2023-29403', 'CVE-2023-39533', 'CVE-2023-39322', 'CVE-2023-39321']
```

File: bin/twistcli_validator.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+            print("vulnerabilities is in ignored cve list and are tracked until fixed")
+        else:
+            print("New vulnerabilities found.Fix them or add into cve ignore list")
+            print([cve for cve in cve_list if cve not in ignored_cve_list])
+            sys.exit(1)
+except FileNotFoundError:
+    if cve_list:
+        print("New vulnerabilities found.Fix them or add into ignore list")
+        print([cve for cve in cve_list if cve not in ignored_cve_list])
+        sys.exit(1)
```
Comment (danielhoherd) [inline bin/twistcli_validator.py:34]: I refactored this whole script:
```suggestion
#!/usr/bin/env python3
import json
from pathlib import Path
import os


git_root = [x for x in Path(__file__).resolve().parents if (x / ".git").is_dir()][
    -1
]
project_name = os.getenv("PROJECT_DIRECTORY")
scan_results_file_path = f"{git_root}/{project_name}"

data = json.loads(Path(f"{scan_results_file_path}/scan-results.json").read_text())
found_cves = [
    vuln["id"] for vuln in data.get("results", [])[0].get("vulnerabilities", [])
]

def cve_list_to_string(list):
    if list:
        return '\n' + '\n'.join([f'  - {item}' for item in list]) + '\n'
    return 'none\n'

def get_ignored_cves():
    ignore_file = Path(f"{scan_results_file_path}/twistcliignore")
    if ignore_file.is_file():
        return [line.strip() for line in ignore_file.read_text().splitlines()]
    return []

ignored_cve_list = get_ignored_cves()

if found_cves:
    print(f"New vulnerabilities found. Fix them or add them to {project_name}/twistcliignore\n")

    new_cves = [cve for cve in found_cves if cve not in ignored_cve_list]
    print(f"New CVEs that are not found in {project_name}/twistcliignore: {cve_list_to_string(new_cves)}")

    old_cves = [cve for cve in found_cves if cve in ignored_cve_list and cve in found_cves]
    print(f"Old CVEs that exist now but are already in {project_name}/twistcliignore: {cve_list_to_string(old_cves)}")

    solved_cves = [cve for cve in ignored_cve_list if cve not in found_cves]
    print(f"Solved CVEs that can be removed from {project_name}/twistcliignore: {cve_list_to_string(solved_cves)}")

    raise SystemExit(1)
```

The output looks like this:
```
$ bin/twistcli_validator.py
2023-09-26T14:17:53-0400
New vulnerabilities found. Fix them or add them to pgbouncer-exporter/twistcliignore

New CVEs that are not found in pgbouncer-exporter/twistcliignore:
  - CVE-2023-29405
  - CVE-2023-29404
  - CVE-2023-29402

Old CVEs that exist now but are already in pgbouncer-exporter/twistcliignore:
  - CVE-2023-39320

Solved CVEs that can be removed from pgbouncer-exporter/twistcliignore:
  - CVE-5555

```

**Review by danielhoherd — APPROVED** [review]
I think we're good here. We can clean up anything that slipped through the cracks in the next iteration.

**Pull request merged**